support/signal.c

```diff
@@ -166,6 +167,90 @@ keep_trying (int r)
 	return r == -1 && errno == EINTR;
 }
 
+// This tiny ad-hoc read/write lock is needed because of the very specific
+// synchronization needed between default_handler and teardown_pipes:
+// - Many default_handlers can be running at once
+// - External forces already ensure only one teardown_pipes runs at once
```

Comment:
Might be good to elaborate upon "external forces", e.g. that `teardown_pipes()` is only invoked when the `signals_mutex` mutex is held, so it *can't* be run concurrently.